IP distance is ascending order indexed by RNSG

**Please state you issue using the following template and, most importantly, in English.**

**Describe the bug**
```
self = <test_search.TestSearchBase object at 0x7fae39a62f98>, connect = <milvus.client.stub.Milvus object at 0x7fae3b2f29e8>, ip_collection = 'test_search_h93uJVGO'
get_simple_index = {'index_param': {'candidate_pool_size': 100, 'knng': 50, 'out_degree': 40, 'search_length': 50}, 'index_type': <IndexType: RNSG>}

    @pytest.mark.level(2)
    def test_search_ip_index_params(self, connect, ip_collection, get_simple_index):
        '''
        target: test basic search fuction, all the search params is corrent, test all index params, and build
        method: search with the given vectors, check the result
        expected: search status ok, and the length of the result is top_k
        '''
        top_k = 10
        index_param = get_simple_index["index_param"]
        index_type = get_simple_index["index_type"]
        logging.getLogger().info(get_simple_index)
        vectors, ids = self.init_data(connect, ip_collection)
        status = connect.create_index(ip_collection, index_type, index_param)
        query_vec = [vectors[0]]
        search_param = get_search_param(index_type)
        status, result = connect.search(ip_collection, top_k, query_vec, params=search_param)
        logging.getLogger().info(result)
        assert status.OK()
        assert len(result[0]) == min(len(vectors), top_k)
>       assert check_result(result[0], ids[0])
E       assert False
E        +  where False = check_result(<milvus.client.abstract.RowQueryResult object at 0x7fae39a62eb8>, 1594434623948827000)

test_search.py:418: AssertionError
----------------------------------------------------------------------------------------------------------- Captured log call ------------------------------------------------------------------------------------------------------------
[2020-07-11 10:30:23-INFO-root]: {'index_type': <IndexType: RNSG>, 'index_param': {'search_length': 50, 'out_degree': 40, 'candidate_pool_size': 100, 'knng': 50}} (test_search.py:409)
[2020-07-11 10:30:28-INFO-root]: [
[
(id:1594434623948839450, distance:0.35238778591156006),
(id:1594434623948833155, distance:0.35504645109176636),
(id:1594434623948836474, distance:0.35881727933883667),
(id:1594434623948827133, distance:0.3637646436691284),
(id:1594434623948835154, distance:0.36582472920417786),
(id:1594434623948835165, distance:0.366372287273407),
(id:1594434623948833294, distance:0.36653730273246765),
(id:1594434623948833998, distance:0.3669421374797821),
(id:1594434623948835033, distance:0.3671693503856659),
(id:1594434623948835358, distance:0.3701755702495575)
]
] (test_search.py:415)
```

**Steps/Code to reproduce behavior**
Follow this [guide](http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) to craft a minimal bug report. This helps us reproduce the issue you're having and resolve the issue more quickly.

**Expected behavior**
A clear and concise description of what you expected to happen.

**Environment details**
0.10.1

**Screenshots**
If applicable, add screenshots to help explain your problem.

**Additional context**
Add any other context about the problem here.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -20,6 +20,7 @@ Please mark all change in change log and use the issue from GitHub
 -   \#2752 Milvus formats vectors data to double-precision and return to http client
 -   \#2767 fix a bug of getting wrong nprobe limitation in knowhere on GPU version
 -   \#2776 Fix too many data copies during creating IVF index
+-   \#2813 To implemente RNSG IP
 
 ## Feature
 

--- a/core/src/index/knowhere/knowhere/index/vector_index/IndexNSG.cpp
+++ b/core/src/index/knowhere/knowhere/index/vector_index/IndexNSG.cpp
@@ -140,7 +140,18 @@ NSG::Train(const DatasetPtr& dataset_ptr, const Config& config) {
     b_params.search_length = config[IndexParams::search_length];
 
     GETTENSORWITHIDS(dataset_ptr)
-    index_ = std::make_shared<impl::NsgIndex>(dim, rows, config[Metric::TYPE].get<std::string>());
+
+    impl::NsgIndex::Metric_Type metric;
+    auto metric_str = config[Metric::TYPE].get<std::string>();
+    if (metric_str == knowhere::Metric::IP) {
+        metric = impl::NsgIndex::Metric_Type::Metric_Type_IP;
+    } else if (metric_str == knowhere::Metric::L2) {
+        metric = impl::NsgIndex::Metric_Type::Metric_Type_L2;
+    } else {
+        KNOWHERE_THROW_MSG("Metric is not supported");
+    }
+
+    index_ = std::make_shared<impl::NsgIndex>(dim, rows, metric);
     index_->SetKnnGraph(knng);
     index_->Build_with_ids(rows, (float*)p_data, (int64_t*)p_ids, b_params);
 }

--- a/core/src/index/knowhere/knowhere/index/vector_index/impl/nsg/Distance.cpp
+++ b/core/src/index/knowhere/knowhere/index/vector_index/impl/nsg/Distance.cpp
@@ -237,7 +237,7 @@ DistanceL2::Compare(const float* a, const float* b, unsigned size) const {
 
 float
 DistanceIP::Compare(const float* a, const float* b, unsigned size) const {
-    return faiss::fvec_inner_product(a, b, (size_t)size);
+    return -(faiss::fvec_inner_product(a, b, (size_t)size));
 }
 
 #endif

--- a/core/src/index/knowhere/knowhere/index/vector_index/impl/nsg/NSG.cpp
+++ b/core/src/index/knowhere/knowhere/index/vector_index/impl/nsg/NSG.cpp
@@ -31,11 +31,11 @@ namespace impl {
 
 unsigned int seed = 100;
 
-NsgIndex::NsgIndex(const size_t& dimension, const size_t& n, std::string metric)
+NsgIndex::NsgIndex(const size_t& dimension, const size_t& n, Metric_Type metric)
     : dimension(dimension), ntotal(n), metric_type(metric) {
-    if (metric == knowhere::Metric::L2) {
+    if (metric == Metric_Type::Metric_Type_L2) {
         distance_ = new DistanceL2;
-    } else if (metric == knowhere::Metric::IP) {
+    } else if (metric == Metric_Type::Metric_Type_IP) {
         distance_ = new DistanceIP;
     }
 }
@@ -406,7 +406,6 @@ NsgIndex::GetNeighbors(const float* query, std::vector<Neighbor>& resset, Graph&
                     // std::cout << "pos: " << pos << ", nn: " << nn.id << ":" << nn.distance << ", nup: " <<
                     // nearest_updated_pos << std::endl;
                     /////
-
                     // trick: avoid search query search_length < init_ids.size() ...
                     if (buffer_size + 1 < resset.size())
                         ++buffer_size;
@@ -846,6 +845,8 @@ NsgIndex::Search(const float* query, const unsigned& nq, const unsigned& dim, co
         }
     }
     rc.RecordSection("search");
+
+    bool is_ip = (metric_type == Metric_Type::Metric_Type_IP);
     for (unsigned int i = 0; i < nq; ++i) {
         unsigned int pos = 0;
         for (unsigned int j = 0; j < resset[i].size(); ++j) {
@@ -853,7 +854,7 @@ NsgIndex::Search(const float* query, const unsigned& nq, const unsigned& dim, co
                 break;  // already top k
             if (!bitset || !bitset->test((faiss::ConcurrentBitset::id_type_t)resset[i][j].id)) {
                 ids[i * k + pos] = ids_[resset[i][j].id];
-                dist[i * k + pos] = resset[i][j].distance;
+                dist[i * k + pos] = is_ip ? -resset[i][j].distance : resset[i][j].distance;
                 ++pos;
             }
         }

--- a/core/src/index/knowhere/knowhere/index/vector_index/impl/nsg/NSG.h
+++ b/core/src/index/knowhere/knowhere/index/vector_index/impl/nsg/NSG.h
@@ -43,9 +43,14 @@ using Graph = std::vector<std::vector<node_t>>;
 
 class NsgIndex {
  public:
+    enum Metric_Type {
+        Metric_Type_L2,
+        Metric_Type_IP,
+    };
+
     size_t dimension;
-    size_t ntotal;            // totabl nb of indexed vectors
-    std::string metric_type;  // L2 | IP
+    size_t ntotal;        // totabl nb of indexed vectors
+    int32_t metric_type;  // enum Metric_Type
     Distance* distance_;
 
     float* ori_data_;
@@ -65,7 +70,7 @@ class NsgIndex {
     size_t out_degree;
 
  public:
-    explicit NsgIndex(const size_t& dimension, const size_t& n, std::string metric = knowhere::Metric::L2);
+    explicit NsgIndex(const size_t& dimension, const size_t& n, Metric_Type metric);
 
     NsgIndex() = default;
 

--- a/core/src/index/knowhere/knowhere/index/vector_index/impl/nsg/NSGIO.cpp
+++ b/core/src/index/knowhere/knowhere/index/vector_index/impl/nsg/NSGIO.cpp
@@ -19,6 +19,7 @@ namespace impl {
 
 void
 write_index(NsgIndex* index, MemoryIOWriter& writer) {
+    writer(&index->metric_type, sizeof(int32_t), 1);
     writer(&index->ntotal, sizeof(index->ntotal), 1);
     writer(&index->dimension, sizeof(index->dimension), 1);
     writer(&index->navigation_point, sizeof(index->navigation_point), 1);
@@ -36,9 +37,11 @@ NsgIndex*
 read_index(MemoryIOReader& reader) {
     size_t ntotal;
     size_t dimension;
+    int32_t metric;
+    reader(&metric, sizeof(int32_t), 1);
     reader(&ntotal, sizeof(size_t), 1);
     reader(&dimension, sizeof(size_t), 1);
-    auto index = new NsgIndex(dimension, ntotal);
+    auto index = new NsgIndex(dimension, ntotal, (impl::NsgIndex::Metric_Type)metric);
     reader(&index->navigation_point, sizeof(index->navigation_point), 1);
 
     index->ori_data_ = new float[index->ntotal * index->dimension];

